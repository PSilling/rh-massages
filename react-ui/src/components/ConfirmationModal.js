// react imports
import React, { Component } from 'react';

// component imports
import ModalActions from './ModalActions';

// module imports
import {ModalContainer, ModalDialog} from 'react-modal-dialog';

// util imports
import _t from '../utils/Translations.js';

/**
 * Custom confirmation modal component.
 */
class ConfirmationModal extends Component {

  render() {
    return(
      <ModalContainer onClose={this.props.onClose}>
        <ModalDialog onClose={this.props.onClose} width="40%">
            <h3>{ _t.translate('Action confirmation') }</h3>
            <p>{this.props.message}</p>
            <ModalActions
              primaryLabel={ _t.translate('Proceed') }
              onProceed={this.props.onConfirm}
              onClose={this.props.onClose}
            />
          </ModalDialog>
      </ModalContainer>
    )
  }
}

export default ConfirmationModal
