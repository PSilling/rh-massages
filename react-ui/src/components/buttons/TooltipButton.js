// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import { Button, Tooltip } from "reactstrap";

// util imports
import Util from "../../util/Util";

/**
 * An outline button component with an attached tooltip.
 */
class TooltipButton extends Component {
  state = { tooltipActive: false };

  tooltipTarget = Util.getTooltipTargets(1)[0];

  toggleTooltip = () => {
    this.setState(prevState => ({ tooltipActive: !prevState.tooltipActive }));
  };

  render() {
    const { disabled, label, tooltip, ...rest } = this.props;
    return (
      <span className="no-print">
        <Button {...rest} id={this.tooltipTarget} outline disabled={this.props.disabled}>
          {this.props.label}
        </Button>
        {this.props.tooltip !== "" && !this.props.disabled && (
          <Tooltip
            isOpen={this.state.tooltipActive}
            target={this.tooltipTarget}
            toggle={this.toggleTooltip}
            trigger="hover"
          >
            {this.props.tooltip}
          </Tooltip>
        )}
      </span>
    );
  }
}

TooltipButton.propTypes = {
  /** function to be called on button click */
  onClick: PropTypes.func.isRequired,
  /** whether the button should be displayed as active or not */
  active: PropTypes.bool,
  /** whether the button should be disabled */
  disabled: PropTypes.bool,
  /** button label to be displayed */
  label: PropTypes.string,
  /** tooltip shown over the button */
  tooltip: PropTypes.string
};

TooltipButton.defaultProps = {
  active: false,
  disabled: false,
  label: "Click me",
  tooltip: ""
};

export default TooltipButton;
