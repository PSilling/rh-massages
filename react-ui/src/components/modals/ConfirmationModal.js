// react imports
import React, { Component } from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";

// module imports
import { ModalContainer, ModalDialog } from "react-modal-dialog";

// component imports
import ModalActions from "../buttons/ModalActions";

// util imports
import _t from "../../util/Translations";

/**
 * Modal with a simple action confirmation message and ModalActions.
 */
class ConfirmationModal extends Component {
  handleKeyPress = event => {
    if (event.charCode === 13 && document.activeElement === ReactDOM.findDOMNode(this.modal)) {
      this.props.onConfirm();
    }
  };

  renderInsides = () => (
    <div>
      <h3>{_t.translate("Action confirmation")}</h3>
      <p>{this.props.message}</p>
      <ModalActions onProceed={this.props.onConfirm} onClose={this.props.onClose} autoFocus>
        {this.props.children}
      </ModalActions>
    </div>
  );

  render() {
    return this.props.withPortal ? (
      <ModalContainer onClose={this.props.onClose}>
        <ModalDialog
          onClose={this.props.onClose}
          width="40%"
          style={{ outline: "none" }}
          tabIndex="-1"
          onKeyPress={this.handleKeyPress}
          ref={dialog => {
            this.modalDialog = dialog;
          }}
        >
          {this.renderInsides()}
        </ModalDialog>
      </ModalContainer>
    ) : (
      this.renderInsides()
    );
  }
}

ConfirmationModal.propTypes = {
  /** children supplied to ModalActions */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  /** message in the Modal */
  message: PropTypes.string.isRequired,
  /** callback function triggered on Modal close */
  onClose: PropTypes.func.isRequired,
  /** callback function triggered on primary button click */
  onConfirm: PropTypes.func.isRequired,
  /** whether ModalContainer should be used; useful for testing to avoid portals */
  withPortal: PropTypes.bool
};

ConfirmationModal.defaultProps = {
  children: null,
  withPortal: true
};

export default ConfirmationModal;
