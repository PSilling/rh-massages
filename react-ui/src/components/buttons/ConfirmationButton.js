// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// component imports
import ConfirmationModal from "../modals/ConfirmationModal";
import TooltipButton from "./TooltipButton";

// util imports
import _t from "../../util/Translations";

/**
 * A TooltipButton with a ConfirmationModal beside it.
 */
class ConfirmationButton extends Component {
  state = { active: false };

  handleToggle = () => {
    this.setState(prevState => ({ active: !prevState.active }));
  };

  render() {
    const { onConfirm, dialogMessage, ...rest } = this.props;
    return (
      <span>
        <TooltipButton {...rest} onClick={this.handleToggle} />
        {this.state.active && (
          <ConfirmationModal
            message={this.props.dialogMessage}
            onClose={this.handleToggle}
            onConfirm={() => {
              this.handleToggle();
              this.props.onConfirm();
            }}
          />
        )}
      </span>
    );
  }
}

ConfirmationButton.propTypes = {
  /** function to be called on action confirmation */
  onConfirm: PropTypes.func.isRequired,
  /** message disabled in the ConfirmationModal */
  dialogMessage: PropTypes.string,
  /** whether the button should be disabled */
  disabled: PropTypes.bool,
  /** button label to be displayed */
  label: PropTypes.string,
  /** tooltip shown over the button */
  tooltip: PropTypes.string
};

ConfirmationButton.defaultProps = {
  dialogMessage: _t.translate("Are you sure? This action cannot be reverted."),
  disabled: false,
  label: "Delete",
  tooltip: "Delete"
};

export default ConfirmationButton;
