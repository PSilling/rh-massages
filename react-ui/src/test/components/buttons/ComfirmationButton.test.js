// react imports
import React from "react";
import { shallow } from "enzyme";

// test imports
import ConfirmationButton from "../../../components/buttons/ConfirmationButton";
import TooltipButton from "../../../components/buttons/TooltipButton";
import ConfirmationModal from "../../../components/modals/ConfirmationModal";

test("renders content with correct props", () => {
  const testFunction = jest.fn();
  const wrapper = shallow(<ConfirmationButton onConfirm={testFunction} label="test" dialogMessage="message" />);
  const button = wrapper.find(TooltipButton);

  button.props().onClick();

  expect(testFunction).not.toHaveBeenCalled();
  expect(wrapper.instance().state.active).toBe(true);

  const dialog = wrapper.find(ConfirmationModal);
  dialog.props().onConfirm();

  expect(wrapper.instance().state.active).toBe(false);
  expect(testFunction).toHaveBeenCalledTimes(1);
  expect(button.props().label).toEqual("test");
  expect(button.props().tooltip).toEqual("Delete");
  expect(button.props().disabled).toBe(false);
  expect(dialog.props().message).toEqual("message");
  expect(wrapper).toMatchSnapshot();
});
