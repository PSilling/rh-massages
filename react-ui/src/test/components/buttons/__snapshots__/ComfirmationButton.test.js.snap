// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content with correct props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ConfirmationButton
    dialogMessage="message"
    disabled={false}
    label="test"
    onConfirm={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    tooltip="Delete"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <TooltipButton
          active={false}
          disabled={false}
          label="test"
          onClick={[Function]}
          tooltip="Delete"
        />,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "active": false,
          "disabled": false,
          "label": "test",
          "onClick": [Function],
          "tooltip": "Delete",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <TooltipButton
            active={false}
            disabled={false}
            label="test"
            onClick={[Function]}
            tooltip="Delete"
          />,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "disabled": false,
            "label": "test",
            "onClick": [Function],
            "tooltip": "Delete",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
